File: sf_trading/sf_trading/page/payment_advice_builder/payment_advice_builder.js
// apps/sf_trading/sf_trading/sf_trading/page/payment_advice_builder/payment_advice_builder.js
// Payment Advice Builder — sweep outstanding vouchers, group them by party, tick what to
// pay, and raise one Payment Advice per party in a single action.
//
// Every figure shown here comes from the server (ERPNext's Payment Entry outstanding
// engine), so what you tick is what gets allocated. Nothing is submitted: advices are
// created as drafts unless "Submit created advices" is ticked, and submission still obeys
// the advice's approver rule.

frappe.pages["payment-advice-builder"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Payment Advice Builder"),
		single_column: true,
	});
	new PaymentAdviceBuilder(page);
};

class PaymentAdviceBuilder {
	constructor(page) {
		this.page = page;
		this.data = null;
		this.selected = new Map(); // party -> Set(reference_record)
		this.make_filters();
		this.make_actions();
		this.body = $('<div class="pab-body">').appendTo(this.page.main);
		this.render_empty(__("Set your filters and hit Fetch Outstanding."));
	}

	// ── filters ──────────────────────────────────────────────────────────────────
	make_filters() {
		this.filters = {};
		const add = (df) => {
			this.filters[df.fieldname] = this.page.add_field(df);
		};

		add({
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
			reqd: 1,
			change: () => this.render_empty(__("Filters changed — fetch again.")),
		});
		add({
			fieldname: "party_type",
			label: __("Party Type"),
			fieldtype: "Select",
			options: ["Supplier", "Customer"].join("\n"),
			default: "Supplier",
		});
		add({ fieldname: "party", label: __("Party"), fieldtype: "Dynamic Link", options: "party_type" });
		add({
			fieldname: "due_before",
			label: __("Due On or Before"),
			fieldtype: "Date",
			default: frappe.datetime.get_today(),
		});
		add({ fieldname: "min_ageing", label: __("Ageing Over (days)"), fieldtype: "Int" });
		add({
			fieldname: "minimum_total",
			label: __("Min Advice Total"),
			fieldtype: "Currency",
			default: 1,
			description: __("Parties whose total falls below this are skipped as rounding residue."),
		});
		add({
			fieldname: "cost_center",
			label: __("Cost Center"),
			fieldtype: "Link",
			options: "Cost Center",
		});
		add({ fieldname: "branch", label: __("Branch"), fieldtype: "Link", options: "Branch" });
		add({ fieldname: "ignore_on_hold", label: __("Include On-Hold Parties"), fieldtype: "Check" });
	}

	get_filters() {
		const values = {};
		Object.keys(this.filters).forEach((key) => {
			const value = this.filters[key].get_value();
			if (value) values[key] = value;
		});
		return values;
	}

	// ── toolbar ──────────────────────────────────────────────────────────────────
	make_actions() {
		this.page.set_primary_action(__("Fetch Outstanding"), () => this.fetch(), "refresh");
		this.page.add_menu_item(__("Select All Payable"), () => this.select_all(true));
		this.page.add_menu_item(__("Clear Selection"), () => this.select_all(false));
	}

	// ── data ─────────────────────────────────────────────────────────────────────
	fetch() {
		const filters = this.get_filters();
		if (!filters.company) {
			frappe.msgprint(__("Pick a company first."));
			return;
		}

		frappe.call({
			method: "sf_trading.api.payment_advice_builder.get_builder_data",
			args: { filters: filters },
			freeze: true,
			freeze_message: __("Sweeping outstanding vouchers…"),
			callback: (r) => {
				this.data = r.message;
				this.selected.clear();
				this.render();
			},
		});
	}

	select_all(on) {
		if (!this.data) return;
		this.data.groups.forEach((group) => {
			if (group.skip) return;
			if (on) {
				this.selected.set(group.party, new Set(group.rows.map((row) => row.reference_record)));
			} else {
				this.selected.delete(group.party);
			}
		});
		this.render();
	}

	// ── render ───────────────────────────────────────────────────────────────────
	render_empty(message) {
		this.body.html(
			`<div class="text-muted" style="padding:2rem;text-align:center">${frappe.utils.escape_html(
				message
			)}</div>`
		);
	}

	render() {
		if (!this.data || !this.data.groups.length) {
			this.render_empty(__("Nothing outstanding for these filters."));
			return;
		}

		const payable = this.data.groups.filter((g) => !g.skip);
		const skipped = this.data.groups.filter((g) => g.skip);

		this.body.empty();
		this.render_summary(payable, skipped);
		payable.forEach((group) => this.render_group(group));
		if (skipped.length) this.render_skipped(skipped);
		this.render_footer();
		this.bind_events();
		this.update_totals();
	}

	render_summary(payable, skipped) {
		$(`
			<div class="pab-summary" style="display:flex;gap:2rem;flex-wrap:wrap;padding:1rem 0;border-bottom:1px solid var(--border-color)">
				<div><div class="text-muted small">${__("Parties")}</div><div class="h5">${payable.length}</div></div>
				<div><div class="text-muted small">${__("Vouchers")}</div><div class="h5">${this.data.totals.vouchers}</div></div>
				<div><div class="text-muted small">${__("Outstanding")}</div><div class="h5">${format_currency(
			this.data.totals.outstanding,
			this.data.currency
		)}</div></div>
				<div><div class="text-muted small">${__("Skipped")}</div><div class="h5">${skipped.length}</div></div>
			</div>
		`).appendTo(this.body);
	}

	render_group(group) {
		const rows = group.rows
			.map(
				(row) => `
			<tr>
				<td><input type="checkbox" class="pab-row" data-party="${frappe.utils.escape_html(
					group.party
				)}" data-ref="${frappe.utils.escape_html(row.reference_record)}"></td>
				<td>${frappe.utils.escape_html(row.reference_doctype)}</td>
				<td><a href="/app/${frappe.router.slug(row.reference_doctype)}/${encodeURIComponent(
					row.reference_record
				)}" target="_blank">${frappe.utils.escape_html(row.reference_record)}</a></td>
				<td>${frappe.utils.escape_html(row.bill_no || "")}</td>
				<td>${frappe.datetime.str_to_user(row.date) || ""}</td>
				<td class="text-right ${row.ageing > 90 ? "text-danger" : ""}">${row.ageing}</td>
				<td>${frappe.utils.escape_html(row.reference_status || "")}</td>
				<td class="text-right">${format_currency(row.net_payable_amount, row.currency)}</td>
			</tr>`
			)
			.join("");

		$(`
			<div class="pab-group" style="margin:1rem 0;border:1px solid var(--border-color);border-radius:var(--border-radius-md)">
				<div style="display:flex;align-items:center;gap:.75rem;padding:.75rem 1rem;background:var(--fg-color)">
					<input type="checkbox" class="pab-group-check" data-party="${frappe.utils.escape_html(group.party)}">
					<b>${frappe.utils.escape_html(group.party_name)}</b>
					<span class="text-muted small">${frappe.utils.escape_html(group.party)}</span>
					${group.on_hold ? `<span class="indicator-pill orange">${__("On Hold")}</span>` : ""}
					${
						group.bank_account
							? ""
							: `<span class="indicator-pill gray" title="${__(
									"No party bank account on file"
							  )}">${__("No bank account")}</span>`
					}
					<span style="margin-left:auto" class="text-muted small">${group.voucher_count} ${__(
			"voucher(s)"
		)} · ${__("oldest")} ${group.oldest_ageing}${__("d")}</span>
					<b>${format_currency(group.total_outstanding, this.data.currency)}</b>
				</div>
				<table class="table table-sm" style="margin:0">
					<thead><tr class="text-muted small">
						<th style="width:32px"></th><th>${__("Type")}</th><th>${__("Document")}</th>
						<th>${__("Party Doc No")}</th><th>${__("Date")}</th>
						<th class="text-right">${__("Ageing")}</th><th>${__("Status")}</th>
						<th class="text-right">${__("Outstanding")}</th>
					</tr></thead>
					<tbody>${rows}</tbody>
				</table>
			</div>
		`).appendTo(this.body);
	}

	render_skipped(skipped) {
		const items = skipped
			.map(
				(group) =>
					`<li>${frappe.utils.escape_html(group.party_name)} — <span class="text-muted">${
						frappe.utils.escape_html(group.skip_label || group.skip)
					}</span></li>`
			)
			.join("");
		$(`
			<div style="margin:1.5rem 0;padding:1rem;border:1px dashed var(--border-color);border-radius:var(--border-radius-md)">
				<div class="text-muted" style="margin-bottom:.5rem">${__("Skipped")} (${skipped.length})</div>
				<ul style="margin:0;padding-left:1.2rem">${items}</ul>
			</div>
		`).appendTo(this.body);
	}

	render_footer() {
		this.footer = $(`
			<div class="pab-footer" style="position:sticky;bottom:0;background:var(--fg-color);border-top:1px solid var(--border-color);padding:1rem;display:flex;align-items:center;gap:1rem">
				<div class="pab-selection text-muted"></div>
				<div style="margin-left:auto;display:flex;gap:.5rem">
					<button class="btn btn-default btn-sm pab-options">${__("Advice Options")}</button>
					<button class="btn btn-primary btn-sm pab-create" disabled>${__("Create Advices")}</button>
				</div>
			</div>
		`).appendTo(this.body);
		this.options = {};
	}

	bind_events() {
		this.body.find(".pab-group-check").on("change", (e) => {
			const party = $(e.currentTarget).data("party");
			const on = e.currentTarget.checked;
			const group = this.data.groups.find((g) => g.party === party);
			if (on) {
				this.selected.set(party, new Set(group.rows.map((r) => r.reference_record)));
			} else {
				this.selected.delete(party);
			}
			this.body
				.find(`.pab-row[data-party="${party}"]`)
				.prop("checked", on);
			this.update_totals();
		});

		this.body.find(".pab-row").on("change", (e) => {
			const el = $(e.currentTarget);
			const party = el.data("party");
			const ref = el.data("ref");
			if (!this.selected.has(party)) this.selected.set(party, new Set());
			const set = this.selected.get(party);
			e.currentTarget.checked ? set.add(ref) : set.delete(ref);
			if (!set.size) this.selected.delete(party);
			this.update_totals();
		});

		this.body.find(".pab-options").on("click", () => this.edit_options());
		this.body.find(".pab-create").on("click", () => this.create());
	}

	update_totals() {
		let parties = 0;
		let vouchers = 0;
		let amount = 0;

		this.selected.forEach((refs, party) => {
			const group = this.data.groups.find((g) => g.party === party);
			if (!group) return;
			parties += 1;
			group.rows.forEach((row) => {
				if (refs.has(row.reference_record)) {
					vouchers += 1;
					amount += flt(row.net_payable_amount);
				}
			});
		});

		this.body
			.find(".pab-selection")
			.html(
				parties
					? __("{0} party(ies), {1} voucher(s) — {2}", [
							parties,
							vouchers,
							`<b>${format_currency(amount, this.data.currency)}</b>`,
					  ])
					: __("Nothing selected")
			);
		this.body.find(".pab-create").prop("disabled", !parties);
	}

	edit_options() {
		const d = new frappe.ui.Dialog({
			title: __("Advice Options"),
			fields: [
				{
					fieldname: "mode_of_payment",
					label: __("Mode of Payment"),
					fieldtype: "Link",
					options: "Mode of Payment",
					default: this.options.mode_of_payment,
				},
				{
					fieldname: "bank_account",
					label: __("Company Bank Account"),
					fieldtype: "Link",
					options: "Bank Account",
					default: this.options.bank_account,
					get_query: () => ({
						filters: {
							company: this.get_filters().company,
							is_company_account: 1,
						},
					}),
				},
				{
					fieldname: "approver",
					label: __("Approver"),
					fieldtype: "Link",
					options: "Employee",
					default: this.options.approver,
				},
				{
					fieldname: "cost_center",
					label: __("Cost Center"),
					fieldtype: "Link",
					options: "Cost Center",
					default: this.options.cost_center,
				},
				{ fieldname: "remarks", label: __("Remarks"), fieldtype: "Small Text", default: this.options.remarks },
				{
					fieldname: "submit",
					label: __("Submit created advices"),
					fieldtype: "Check",
					description: __("Submission still requires you to be the selected Approver."),
					default: this.options.submit,
				},
			],
			primary_action_label: __("Save Options"),
			primary_action: (values) => {
				this.options = values;
				d.hide();
				frappe.show_alert({ message: __("Options saved"), indicator: "green" }, 3);
			},
		});
		d.show();
	}

	create() {
		const filters = this.get_filters();
		const selections = [];

		this.selected.forEach((refs, party) => {
			const group = this.data.groups.find((g) => g.party === party);
			if (!group) return;
			const references = group.rows.filter((row) => refs.has(row.reference_record));
			if (references.length) {
				selections.push({ party: party, references: references, bank_account: group.bank_account });
			}
		});

		const total = selections.reduce(
			(sum, s) => sum + s.references.reduce((rs, r) => rs + flt(r.net_payable_amount), 0),
			0
		);

		frappe.confirm(
			__("Create {0} Payment Advice(s) totalling {1}?", [
				selections.length,
				format_currency(total, this.data.currency),
			]),
			() => {
				frappe.call({
					method: "sf_trading.api.payment_advice_builder.create_advices",
					args: {
						selections: selections,
						options: Object.assign(
							{
								company: filters.company,
								party_type: filters.party_type || "Supplier",
							},
							this.options || {}
						),
					},
					freeze: true,
					freeze_message: __("Creating Payment Advices…"),
					callback: (r) => this.show_result(r.message),
				});
			}
		);
	}

	show_result(result) {
		if (!result) return;

		if (result.queued) {
			frappe.msgprint({ title: __("Queued"), message: result.message, indicator: "blue" });
			return;
		}

		const created = (result.created || [])
			.map(
				(row) =>
					`<li><a href="/app/payment-advice/${encodeURIComponent(row.advice)}" target="_blank">${
						row.advice
					}</a> — ${frappe.utils.escape_html(row.party)} — ${format_currency(row.amount)}</li>`
			)
			.join("");
		const failed = (result.failed || [])
			.map(
				(row) =>
					`<li>${frappe.utils.escape_html(row.party)} — <span class="text-danger">${
						frappe.utils.escape_html(row.error)
					}</span></li>`
			)
			.join("");

		frappe.msgprint({
			title: __("Payment Advices"),
			indicator: failed ? "orange" : "green",
			message: `
				<p>${__("Created")}: <b>${(result.created || []).length}</b> — ${format_currency(
				result.total_amount
			)}</p>
				${created ? `<ul>${created}</ul>` : ""}
				${failed ? `<p>${__("Failed")}:</p><ul>${failed}</ul>` : ""}
			`,
		});

		this.fetch(); // re-sweep so what was just advised drops out of the list
	}
}

File: sf_trading/api/payment_advice_builder.py
# apps/sf_trading/sf_trading/api/payment_advice_builder.py
"""Supplier-wise Payment Advice generator.

One filtered sweep of outstanding vouchers, grouped by party, turned into **one Payment
Advice per party**. This is the piece ERPNext's Payment Request cannot do: a PR is
one-document-per-request, so "pay these nine invoices for this supplier" has no home.

Both entry points share this module, and so will the automation in Phase 3 — the scheduler
calls `create_advices()` with the same payload the builder page sends, so automatic and
manual runs cannot drift apart:

  * the Payment Advice Builder desk page (preview, tick, create)
  * the Purchase Invoice list bulk action

Outstanding amounts come from `shape_reference_rows()`, which wraps ERPNext's own Payment
Entry engine — Payment-Ledger truthful, so part payments, credit notes and return invoices
are already netted.

Nothing here submits anything. Advices are created as drafts unless the caller explicitly
asks to submit, and submission still obeys the advice's approver rule.
"""

import json

import frappe
from frappe import _
from frappe.utils import add_days, cint, flt, getdate, nowdate

from sf_trading.sf_trading.doctype.payment_advice.payment_advice import (
    PAY_PARTY_TYPES,
    RECEIVE_PARTY_TYPES,
    get_party_account,
    shape_reference_rows,
)

# Skip reasons, kept as constants so the page and the tests speak the same language.
SKIP_NO_PARTY_ACCOUNT = "no_party_account"
SKIP_ALREADY_ADVISED = "already_advised"
SKIP_BELOW_FLOOR = "below_floor"
SKIP_ON_HOLD = "supplier_on_hold"
SKIP_DISABLED = "supplier_disabled"
SKIP_NO_ROWS = "nothing_outstanding"

SKIP_LABELS = {
    SKIP_NO_PARTY_ACCOUNT: _("No default payable/receivable account"),
    SKIP_ALREADY_ADVISED: _("Already allocated on a live Payment Advice"),
    SKIP_BELOW_FLOOR: _("Below the minimum advice total"),
    SKIP_ON_HOLD: _("Party is on hold"),
    SKIP_DISABLED: _("Party is disabled"),
    SKIP_NO_ROWS: _("Nothing outstanding after filters"),
}

# Anything at or below this is rounding residue, not a payable. ERPNext will happily return
# a 0.005 outstanding row; raising an advice for half a fils is noise.
DEFAULT_FLOOR = 1.0

ENQUEUE_THRESHOLD = 15  # parties per run before we push the work to a background job


# ── party discovery ──────────────────────────────────────────────────────────────

def _party_field(party_type):
    return {"Supplier": "supplier", "Customer": "customer"}.get(party_type, "supplier")


def _source_doctype(party_type):
    return "Purchase Invoice" if party_type in PAY_PARTY_TYPES else "Sales Invoice"


def _candidate_parties(filters):
    """Parties with something outstanding under the given filters."""
    party_type = filters.get("party_type") or "Supplier"
    doctype = _source_doctype(party_type)
    party_field = _party_field(party_type)

    conditions = {
        "docstatus": 1,
        "company": filters.get("company"),
        "outstanding_amount": [">", 0],
    }

    if filters.get("party"):
        conditions[party_field] = filters.get("party")
    if filters.get("branch"):
        conditions["branch"] = filters.get("branch")
    if filters.get("cost_center"):
        conditions["cost_center"] = filters.get("cost_center")
    if filters.get("currency"):
        conditions["currency"] = filters.get("currency")

    due_before = filters.get("due_before")
    if due_before:
        conditions["due_date"] = ["<=", getdate(due_before)]

    if filters.get("from_posting_date") and filters.get("to_posting_date"):
        conditions["posting_date"] = [
            "between",
            [getdate(filters["from_posting_date"]), getdate(filters["to_posting_date"])],
        ]

    parties = frappe.get_all(
        doctype,
        filters=conditions,
        fields=[party_field + " as party"],
        group_by=party_field,
        order_by=party_field,
        pluck="party",
    )

    if filters.get("party_group"):
        group_field = "supplier_group" if party_type == "Supplier" else "customer_group"
        in_group = frappe.get_all(
            party_type,
            filters={group_field: filters["party_group"], "name": ["in", parties]},
            pluck="name",
        )
        parties = [p for p in parties if p in in_group]

    return parties


def _party_state(party_type, party):
    """Hold / disabled state, so the preview can explain a skip instead of hiding it."""
    state = frappe._dict(on_hold=False, disabled=False, release_date=None)

    if party_type == "Supplier":
        values = frappe.db.get_value(
            "Supplier", party, ["on_hold", "hold_type", "release_date", "disabled"], as_dict=True
        ) or frappe._dict()
        release = values.get("release_date")
        # ERPNext holds a supplier until release_date; a past release date is not a hold
        holding = cint(values.get("on_hold")) and (
            not release or getdate(release) >= getdate(nowdate())
        )
        state.on_hold = bool(holding) and values.get("hold_type") in (None, "", "All", "Payments")
        state.disabled = bool(cint(values.get("disabled")))
        state.release_date = release
    elif party_type == "Customer":
        state.disabled = bool(cint(frappe.db.get_value("Customer", party, "disabled")))

    return state


def _already_advised(records):
    """Vouchers already carrying an allocation on a live (non-cancelled) advice."""
    if not records:
        return set()

    rows = frappe.get_all(
        "Payment Advice Reference",
        filters={
            "reference_record": ["in", records],
            "parenttype": "Payment Advice",
            "docstatus": ["!=", 2],
            "allocated_amount": [">", 0],
        },
        fields=["reference_record", "parent"],
    )
    return {r.reference_record for r in rows}


# ── preview ──────────────────────────────────────────────────────────────────────

@frappe.whitelist()
def get_builder_data(filters=None):
    """Grouped preview: one entry per party, with its outstanding vouchers and skip reason."""
    frappe.has_permission("Payment Advice", "create", throw=True)

    filters = frappe._dict(_as_dict(filters))
    if not filters.get("company"):
        frappe.throw(_("Company is required."))

    party_type = filters.get("party_type") or "Supplier"
    if party_type not in PAY_PARTY_TYPES + RECEIVE_PARTY_TYPES:
        frappe.throw(_("Unsupported Party Type: %s") % party_type)

    floor = flt(filters.get("minimum_total") or DEFAULT_FLOOR)
    min_ageing = cint(filters.get("min_ageing"))

    groups = []
    for party in _candidate_parties(filters):
        group = _build_group(party_type, party, filters, floor, min_ageing)
        if group:
            groups.append(group)

    groups.sort(key=lambda g: -flt(g["total_outstanding"]))
    return {
        "party_type": party_type,
        "company": filters.get("company"),
        "currency": frappe.db.get_value("Company", filters.get("company"), "default_currency"),
        "floor": floor,
        "groups": groups,
        "totals": {
            "parties": len([g for g in groups if not g["skip"]]),
            "vouchers": sum(len(g["rows"]) for g in groups if not g["skip"]),
            "outstanding": flt(
                sum(flt(g["total_outstanding"]) for g in groups if not g["skip"]), 3
            ),
            "skipped": len([g for g in groups if g["skip"]]),
        },
        "skip_labels": SKIP_LABELS,
    }


def _build_group(party_type, party, filters, floor, min_ageing):
    state = _party_state(party_type, party)
    party_account = get_party_account(party_type, party, filters.get("company"))

    rows = _outstanding_rows(party_type, party, filters, party_account, min_ageing)
    advised = _already_advised([r["reference_record"] for r in rows])

    fresh = [r for r in rows if r["reference_record"] not in advised]
    total = flt(sum(flt(r["net_payable_amount"]) for r in fresh), 3)

    skip = None
    if not party_account:
        skip = SKIP_NO_PARTY_ACCOUNT
    elif state.disabled:
        skip = SKIP_DISABLED
    elif state.on_hold and not cint(filters.get("ignore_on_hold")):
        skip = SKIP_ON_HOLD
    elif not fresh:
        skip = SKIP_ALREADY_ADVISED if rows else SKIP_NO_ROWS
    elif total < floor:
        skip = SKIP_BELOW_FLOOR

    return {
        "party": party,
        "party_name": frappe.db.get_value(
            party_type, party, "supplier_name" if party_type == "Supplier" else "customer_name"
        )
        or party,
        "party_account": party_account,
        "bank_account": _party_bank_account(party_type, party),
        "rows": fresh,
        "already_advised": sorted(advised),
        "voucher_count": len(fresh),
        "total_outstanding": total,
        "oldest_ageing": max([cint(r["ageing"]) for r in fresh], default=0),
        "currencies": sorted({r["currency"] for r in fresh if r.get("currency")}),
        "on_hold": state.on_hold,
        "release_date": state.release_date,
        "skip": skip,
        "skip_label": SKIP_LABELS.get(skip) if skip else None,
    }


def _outstanding_rows(party_type, party, filters, party_account, min_ageing):
    """Outstanding vouchers for one party, via ERPNext's Payment Entry engine."""
    if not party_account:
        return []

    from erpnext.accounts.doctype.payment_entry.payment_entry import (
        get_outstanding_reference_documents,
    )

    args = {
        "company": filters.get("company"),
        "party_type": party_type,
        "party": party,
        "party_account": party_account,
        "payment_type": "Pay" if party_type in PAY_PARTY_TYPES else "Receive",
        "posting_date": nowdate(),
        "get_outstanding_invoices": True,
        "get_orders_to_be_billed": cint(filters.get("include_orders")),
        "cost_center": filters.get("cost_center"),
        "to_due_date": getdate(filters["due_before"]) if filters.get("due_before") else None,
        "from_posting_date": filters.get("from_posting_date"),
        "to_posting_date": filters.get("to_posting_date"),
    }

    try:
        vouchers = get_outstanding_reference_documents(args) or []
    except Exception:
        # a party-level problem (blocked supplier, missing account currency) must not kill
        # the whole sweep — the group simply reports nothing outstanding
        frappe.log_error(frappe.get_traceback(), "sf_trading: outstanding fetch failed for %s" % party)
        return []

    rows = shape_reference_rows(
        vouchers,
        cost_center=filters.get("cost_center"),
        from_amount=filters.get("from_amount"),
        to_amount=filters.get("to_amount"),
    )

    if min_ageing:
        rows = [r for r in rows if cint(r["ageing"]) >= min_ageing]

    return rows


def _party_bank_account(party_type, party):
    return frappe.db.get_value(
        "Bank Account", {"party_type": party_type, "party": party, "is_default": 1}, "name"
    ) or frappe.db.get_value("Bank Account", {"party_type": party_type, "party": party}, "name")


# ── creation ─────────────────────────────────────────────────────────────────────

@frappe.whitelist()
def create_advices(selections, options=None):
    """Create one Payment Advice per selected party.

    `selections` is a list of {party, references: [...], payment_amount?} — exactly the shape
    `get_builder_data()` returns, so the page can hand back what it was given after the user
    ticks rows. Large batches are pushed to a background job.
    """
    frappe.has_permission("Payment Advice", "create", throw=True)

    selections = _as_list(selections)
    options = frappe._dict(_as_dict(options))

    if not selections:
        frappe.throw(_("Select at least one party."))

    if len(selections) > ENQUEUE_THRESHOLD and not options.get("run_now"):
        job = frappe.enqueue(
            "sf_trading.api.payment_advice_builder.create_advices_bulk",
            queue="long",
            timeout=1800,
            enqueue_after_commit=True,
            job_name="payment_advice_builder_%s" % frappe.session.user,
            selections=selections,
            options=dict(options),
            user=frappe.session.user,
        )
        return {
            "queued": True,
            "job": getattr(job, "id", None),
            "parties": len(selections),
            "message": _("Creating %s advices in the background. You will be notified when it finishes.")
            % len(selections),
        }

    return _create_many(selections, options)


def create_advices_bulk(selections, options=None, user=None):
    """Background entry point — same work, then tell the user it landed."""
    result = _create_many(_as_list(selections), frappe._dict(_as_dict(options)))

    if user:
        created = len(result["created"])
        frappe.publish_realtime(
            event="sf_invoice_overdue_alert",  # reuses the live chime + toast + bell channel
            message={
                "title": _("Payment Advices created"),
                "message": _("%(count)s advice(s) created, %(total)s total")
                % {
                    "count": created,
                    "total": frappe.utils.fmt_money(result["total_amount"], precision=3),
                },
                "count": created,
                "outstanding": result["total_amount"],
                "report": "Payment Advice",
            },
            user=user,
            after_commit=True,
        )
    return result


def _create_many(selections, options):
    created, failed = [], []
    total = 0.0

    for selection in selections:
        selection = frappe._dict(selection)
        try:
            advice = _create_one(selection, options)
            created.append({"advice": advice.name, "party": advice.party, "amount": flt(advice.payment_amount)})
            total += flt(advice.payment_amount)
        except Exception as exc:
            frappe.db.rollback()
            failed.append({"party": selection.get("party"), "error": str(exc)})
            frappe.log_error(frappe.get_traceback(), "sf_trading: advice creation failed")
        else:
            frappe.db.commit()

    return {
        "queued": False,
        "created": created,
        "failed": failed,
        "total_amount": flt(total, 3),
    }


def _create_one(selection, options):
    """Build and insert a single Payment Advice from one party's selected references."""
    references = _as_list(selection.get("references"))
    if not references:
        frappe.throw(_("No references selected for %s") % selection.get("party"))

    party_type = options.get("party_type") or "Supplier"
    company = options.get("company")

    advice = frappe.new_doc("Payment Advice")
    advice.update(
        {
            "company": company,
            "payment_advice_type": "Outward" if party_type in PAY_PARTY_TYPES else "Inward",
            "transaction_date": options.get("transaction_date") or nowdate(),
            "party_type": party_type,
            "party": selection.get("party"),
            "mode_of_payment": options.get("mode_of_payment"),
            "bank_account": options.get("bank_account") or selection.get("bank_account"),
            "cost_center": options.get("cost_center"),
            "project": options.get("project"),
            "approver": options.get("approver"),
            "remarks": options.get("remarks"),
            "auto_generated": cint(options.get("auto_generated")),
        }
    )

    for row in references:
        row = frappe._dict(row)
        advice.append(
            "payment_advice_reference",
            {
                "reference_doctype": row.get("reference_doctype"),
                "reference_record": row.get("reference_record"),
                "bill_no": row.get("bill_no"),
                "date": row.get("date"),
                "amount": flt(row.get("amount")),
                "settled_amount": flt(row.get("settled_amount")),
                "net_payable_amount": flt(row.get("net_payable_amount")),
                "payment_term": row.get("payment_term"),
                "currency": row.get("currency"),
                "exchange_rate": flt(row.get("exchange_rate")) or 1.0,
                "cost_center": row.get("cost_center") or options.get("cost_center"),
            },
        )

    payable = flt(sum(flt(r.net_payable_amount) for r in advice.payment_advice_reference), 3)
    requested = flt(selection.get("payment_amount"))
    advice.payment_amount = min(requested, payable) if requested else payable

    advice.insert()

    if cint(options.get("submit")):
        advice.submit()

    return advice


# ── Purchase Invoice list bulk action ────────────────────────────────────────────

@frappe.whitelist()
def create_advices_from_invoices(invoices, options=None):
    """Group hand-picked invoices by their party and raise one advice each."""
    frappe.has_permission("Payment Advice", "create", throw=True)

    invoices = _as_list(invoices)
    if not invoices:
        frappe.throw(_("Select at least one invoice."))

    options = frappe._dict(_as_dict(options))
    doctype = options.get("doctype") or "Purchase Invoice"
    party_type = "Supplier" if doctype == "Purchase Invoice" else "Customer"
    party_field = _party_field(party_type)

    rows = frappe.get_all(
        doctype,
        filters={"name": ["in", invoices], "docstatus": 1, "outstanding_amount": [">", 0]},
        fields=[
            "name",
            party_field + " as party",
            "company",
            "posting_date",
            "due_date",
            "grand_total",
            "outstanding_amount",
            "currency",
            "conversion_rate",
            "cost_center",
            "status",
            "bill_no" if doctype == "Purchase Invoice" else "name as bill_no",
        ],
    )
    if not rows:
        frappe.throw(_("None of the selected documents have an outstanding amount."))

    companies = {r.company for r in rows}
    if len(companies) > 1:
        frappe.throw(_("Select invoices of a single company."))

    already = _already_advised([r.name for r in rows])
    skipped = [r.name for r in rows if r.name in already]

    grouped = {}
    today = getdate(nowdate())
    for row in rows:
        if row.name in already:
            continue
        grouped.setdefault(row.party, []).append(
            {
                "reference_doctype": doctype,
                "reference_record": row.name,
                "bill_no": row.get("bill_no"),
                "date": row.posting_date,
                "amount": flt(row.grand_total),
                "settled_amount": flt(flt(row.grand_total) - flt(row.outstanding_amount), 3),
                "net_payable_amount": flt(row.outstanding_amount),
                "currency": row.currency,
                "exchange_rate": flt(row.conversion_rate) or 1.0,
                "cost_center": row.cost_center,
                "ageing": max(0, (today - getdate(row.due_date or row.posting_date)).days),
            }
        )

    if not grouped:
        frappe.throw(
            _("Every selected document is already allocated on a live Payment Advice: %s")
            % ", ".join(skipped)
        )

    options.update({"company": companies.pop(), "party_type": party_type})
    result = _create_many(
        [{"party": party, "references": refs} for party, refs in grouped.items()], options
    )
    result["skipped_already_advised"] = skipped
    return result


# ── helpers ──────────────────────────────────────────────────────────────────────

def _as_dict(value):
    if not value:
        return {}
    if isinstance(value, str):
        return json.loads(value)
    return dict(value)


def _as_list(value):
    if not value:
        return []
    if isinstance(value, str):
        return json.loads(value)
    return list(value)


def get_due_cutoff(offset_days=0):
    """Cut-off used by both the builder default and the automation's due_date_offset."""
    return add_days(getdate(nowdate()), cint(offset_days))
